### DISCUSSION TITLE
Make C# String static methods act as extension methods

### DISCUSSION THREAD

**Discussion by JustinMinnaar** [dotnet/csharplang#8363] [Language Ideas]
It would be nice if existing static methods on the String class could be changed to extension methods. 

Then, instead of
```
   if (String.IsNullOrWhitespace(x)) ...
```

You would be able to use
```
    If (x.IsNullOrWhitespace()) ...
```

This could easily be done by modifying the string library and adding the `this` operator to the methods.

**Comment by siegfriedpammer** [comment]
This sounds more like an API change request, which should go into the [runtime](https://github.com/dotnet/runtime) repo.

**Reply by JustinMinnaar** [reply]
Thanks for the advice. I'm still trying to understand where to post suggestions. I was actually trying to figure out how to download the .NET runtime and see if I could make this change, and test it myself. However, I'm stumped on that. I've posted this in another discussion to see what suggestions I get back on it.

**Reply by huoyaoyuan** [reply]
> This could easily be done by modifying the string library and adding the `this` operator to the methods.

No. C# language doesn't allow methods in non-static class to be used as extension methods.

**Reply by JustinMinnaar** [reply]
Then we would have to move the static methods to a String Extensions class or create copies. Not so nice. So possibly the language needs to support adding static methods from a non static class as extension methods.

Thank you for the replies, it is great to have some feedback on the idea, even if it can't be implemented now. In my own project, I just create string extension methods that provided this functionality, but it always like a hack since it was just calling the underlying static methods. I just prefered the readibility and found it easier when typing. Poor excuse, I know.

**Reply by JustinMinnaar** [reply]
```
public static class StringExtensions
{
    public static bool IsNullOrWhiteSpace([NotNullWhen(false)] this string? value)
    {
        if (value == null) return true;

        for (int i = 0; i < value.Length; i++)
        {
            // if (!char.IsWhiteSpace(value[i])) return false;
            if (ch > 32 || ch == '\0') return false;
        }

        return true;
    }
}
```

**Reply by KennethHoff** [reply]
> ```
> public static class StringExtensions
> {
>     public static bool IsNullOrWhiteSpace([NotNullWhen(false)] this string? value)
>     {
>         if (value == null) return true;
> 
>         for (int i = 0; i < value.Length; i++)
>         {
>             // if (!char.IsWhiteSpace(value[i])) return false;
>             if (ch > 32 || ch == '\0') return false;
>         }
> 
>         return true;
>     }
> }
> ```

You could (and probably should) simply call `string.IsNullOrWhitespace` in this implementation. Probably more efficient

**Reply by JustinMinnaar** [reply]
I agree with you. This code is actually from .NET's implementation, and I commented out the line as a thought to rather implement the same logic locally. What I'm wondering is what would be required to make these existing functions which are static and take the string as their parameter, act as extension methods. As huoyaoyuan pointed out though, they are not in static classes, and extension methods require a static class to be implemented. So either the language would have to change, or we would have to copy the code to a static class, in order to not break existing code. I had a look at the current code in .NET core and the static methods are in the String class. Possibily in the future the language could support extension methods without requiring a separate class.

**Reply by HaloFour** [reply]
I doubt there'd be an interest in doing that implicitly as even if the static method accepts the class as the first parameter that doesn't imply that it's intended to be invoked like an instance method.  I think the ability to expand where extensions can be defined is captured in #301, but with the effort to move to explicit extension types it's probably not high on the list of things to do.
